_posts/2021-11-18-keras-updates/keras-updates.Rmd

```diff
@@ -0,0 +1,159 @@
+
+So this is where we complement the above "make it R-like and natural, where possible" with "make it easy to port from Python, where necessary". With the new low-level functionality, you won't have to wait for R wrappers to make use of Python-defined objects. Instead, Python objects may be sub-classed directly from R; and any additional functionality you'd like to add to the subclass is defined in a Python-like syntax. What this means, concretely, is that translating Python code to R has become a lot easier. We'll catch a glimpse of this in the second of our three highlights.
+
+## New in Keras 2.6/7: Three highlights
+
+Among the many new capabilities added in Keras 2.6 and 2.7, we quickly introduce three of the very most important.
+
+-   *Pre-processing layers* significantly help to streamline the training workflow, integrating data manipulation and data augmentation.
+
+-   The ability to subclass Python objects (already alluded to several times) is the new low-level magic that, at the same time, is available to the `keras` user *and* is powering many user-facing enhancements underneath.
+
+-   Recurrent neural network (RNN) layers gain a new cell-level API.
+
+Of these, the first two definitely deserve some deeper treatment; more detailed posts will follow.
+
+### Pre-processing layers
+
+Before the advent of these dedicated layers, pre-processing used to be done as part of the `tfdatasets` pipeline. You would chain operations as required; maybe, integrating random transformations to be applied while training. Depending on what you wanted to achieve, significant programming effort may have ensued.
+
+This is one aspect where the new capabilities can help. Pre-processing layers exist for several types of data, allowing for the usual "data wrangling", as well as data augmentation and feature engineering (as in, hashing categorical data, or vectorizing text).
+
+The mention of text vectorization leads to a second advantage. Unlike, say, a random distortion, vectorization is not something that, once done, may be forgotten about. We don't want to lose the original information, namely, the words. The same happens, for numerical data, with normalization. We need to keep the summary statistics. This means there are two types of pre-processing layers: stateless and stateful ones. The former are part of the training process; the latter are called in advance.
```

Comment:
```suggestion
The mention of text vectorization leads to a second advantage. Unlike, say, a random distortion, vectorization is not something that, may be forgotten about once done. We don't want to lose the original information, namely, the words. The same happens, for numerical data, with normalization. We need to keep the summary statistics. This means there are two types of pre-processing layers: stateless and stateful ones. The former are part of the training process; the latter are called in advance.
```

Reply:
That's another sentence that's pretty comma/clause laden. For a sentence that doesn't contain a list, I think that a *refactor* alarm should go off around five commas. 🚨

Reply:
haha yeah! actually, in your version, i think you left in one that needs to go - so down to 3 :-)))